rocky/rocky.go

```diff
@@ -0,0 +1,271 @@
+}
+
+func WithRetry(retry int) option {
+	return func(opts *options) { opts.retry = retry }
+}
+
+func WithReleases(releases []string) option {
+	return func(opts *options) { opts.releases = releases }
+}
+
+func WithRepos(repos []string) option {
+	return func(opts *options) { opts.repos = repos }
+}
+
+func WithArches(arches []string) option {
+	return func(opts *options) { opts.arches = arches }
+}
+
+type Config struct {
+	*options
+}
+
+func NewConfig(opts ...option) Config {
+	o := &options{
+		url:      urlFormat,
+		dir:      filepath.Join(utils.VulnListDir(), rockyDir),
+		retry:    retry,
+		releases: defaultReleases,
+		repos:    defaultRepos,
+		arches:   defaultArches,
+	}
+	for _, opt := range opts {
+		opt(o)
+	}
+
+	return Config{
+		options: o,
+	}
+}
+
+func (c Config) Update() error {
+	for _, release := range c.releases {
+		for _, repo := range c.repos {
+			for _, arch := range c.arches {
+				log.Printf("Fetching Rocky Linux %s %s %s data...\n", release, repo, arch)
+				if err := c.update(release, repo, arch); err != nil {
+					return xerrors.Errorf("failed to update security advisories of Rocky Linux %s %s %s: %w", release, repo, arch, err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (c Config) update(release, repo, arch string) error {
+	dirPath := filepath.Join(c.dir, release, repo, arch)
+	log.Printf("Remove Rocky Linux %s %s %s directory %s\n", release, repo, arch, dirPath)
+	if err := os.RemoveAll(dirPath); err != nil {
+		return xerrors.Errorf("failed to remove Rocky Linux %s %s %s directory: %w", release, repo, arch, err)
+	}
+	if err := os.MkdirAll(dirPath, os.ModePerm); err != nil {
+		return xerrors.Errorf("failed to mkdir: %w", err)
+	}
+
+	u, err := url.Parse(fmt.Sprintf(c.url, release, repo, arch))
+	if err != nil {
+		return xerrors.Errorf("failed to parse root url: %w", err)
+	}
+	rootPath := u.Path
+	u.Path = path.Join(rootPath, "repodata/repomd.xml")
+	updateInfoPath, err := fetchUpdateInfoPath(u.String())
+	if err != nil {
+		return xerrors.Errorf("failed to fetch updateInfo path from repomd.xml: %w", err)
+	}
+	u.Path = path.Join(rootPath, updateInfoPath)
+	uinfo, err := fetchUpdateInfo(u.String())
+	if err != nil {
+		return xerrors.Errorf("failed to fetch updateInfo: %w", err)
+	}
+
+	secErrata := map[string][]RLSA{}
+	for _, rlsa := range uinfo.RLSAList {
+		if !strings.HasPrefix(rlsa.ID, "RLSA-") {
+			continue
+		}
+
+		issuedDate, err := time.Parse("2006-01-02 15:04:05", rlsa.Issued.Date)
+		if err != nil {
+			return xerrors.Errorf("failed to parse issued date: %w", err)
+		}
+		y := strconv.Itoa(issuedDate.Year())
+		secErrata[y] = append(secErrata[y], rlsa)
+	}
+
+	for year, errata := range secErrata {
+		log.Printf("Write Errata for Rocky Linux %s %s %s %s\n", release, repo, arch, year)
+
+		if err := os.MkdirAll(filepath.Join(dirPath, year), os.ModePerm); err != nil {
+			return xerrors.Errorf("failed to mkdir: %w", err)
+		}
+
+		bar := pb.StartNew(len(errata))
```

Comment:
A progress bar may be unnecessary as it is fast enough.

Reply:
I thought the information in the progress bar would be good, because together with the log, it shows immediately how many items have been inserted into which repository, arch, or year.

Reply:
OK, we can keep it then.